src/java/org/apache/cassandra/index/sai/disk/v1/kdtree/BKDWriter.java

```diff
@@ -0,0 +1,782 @@
+
+            out.writeVInt(scratchOut2.getPosition());
+            out.writeBytes(scratchOut2.getBytes(), 0, scratchOut2.getPosition());
+
+            if (callback != null) callback.writeLeafDocs(rowIDAndIndexes, 0, leafCount);
+
+            writeCommonPrefixLength(scratchOut, commonPrefixLength, leafValues);
+
+            scratchBytesRef1.length = bytesPerValue;
+            scratchBytesRef1.bytes = leafValues;
+
+            IntFunction<BytesRef> packedValues = (i) -> {
+                scratchBytesRef1.offset = bytesPerValue * i;
+                return scratchBytesRef1;
+            };
+            if (DEBUG)
+                valuesInOrderAndBounds(leafCount,
+                                       ArrayUtil.copyOfSubArray(leafValues, 0, bytesPerValue),
+                                       ArrayUtil.copyOfSubArray(leafValues, (leafCount - 1) * bytesPerValue, leafCount * bytesPerValue),
+                                       packedValues,
+                                       leafDocs);
+
+            writeLeafBlockPackedValues(scratchOut, commonPrefixLength, leafCount, packedValues);
+
+            out.writeBytes(scratchOut.getBytes(), 0, scratchOut.getPosition());
+
+            scratchOut.reset();
+        }
+    }
+
+    // TODO: there must be a simpler way?
+    private void rotateToTree(int nodeID, int offset, int count, byte[] index, List<byte[]> leafBlockStartValues)
+    {
+        if (count == 1)
+        {
+            // Leaf index node
+            System.arraycopy(leafBlockStartValues.get(offset), 0, index, nodeID * (1 + bytesPerValue) + 1, bytesPerValue);
+        }
+        else if (count > 1)
+        {
+            // Internal index node: binary partition of count
+            int countAtLevel = 1;
+            int totalCount = 0;
+            while (true)
+            {
+                int countLeft = count - totalCount;
+                if (countLeft <= countAtLevel)
+                {
+                    // This is the last level, possibly partially filled:
+                    int lastLeftCount = Math.min(countAtLevel / 2, countLeft);
+                    assert lastLeftCount >= 0;
+                    int leftHalf = (totalCount - 1) / 2 + lastLeftCount;
+
+                    int rootOffset = offset + leftHalf;
+
+                    System.arraycopy(leafBlockStartValues.get(rootOffset), 0, index, nodeID * (1 + bytesPerValue) + 1, bytesPerValue);
+
+                    // TODO: we could optimize/specialize, when we know it's simply fully balanced binary tree
+                    // under here, to save this while loop on each recursion
+
+                    // Recurse left
+                    rotateToTree(2 * nodeID, offset, leftHalf, index, leafBlockStartValues);
+
+                    // Recurse right
+                    rotateToTree(2 * nodeID + 1, rootOffset + 1, count - leftHalf - 1, index, leafBlockStartValues);
```

Comment:
I think this call is tail-recursive, but trying to do it iteratively is giving me some other weird errors, like the inner `while` loop not being able to terminate w/o an exception. Will come back to this, and the TODOs, once I understand what the point of this method is in the first place.